plan/eliminate_projection.go

```diff
@@ -14,50 +14,257 @@
 			return false
 		}
 	}
 	return true
 }
+
+func resolveColumnAndReplace(origin *expression.Column, replace map[string]*expression.Column) {
+	dst := replace[string(origin.HashCode())]
+	if dst != nil {
+		*origin = *dst
+	}
+}
+
+func resolveExprAndReplace(origin expression.Expression, replace map[string]*expression.Column) {
+	switch expr := origin.(type) {
+	case *expression.Column:
+		resolveColumnAndReplace(expr, replace)
+	case *expression.ScalarFunction:
+		for _, arg := range expr.GetArgs() {
+			resolveExprAndReplace(arg, replace)
+		}
+	}
+}
+
+// eliminatePhysicalProjection should be called after physical optimization to eliminate the redundant projection
+// left after logical projection elimination.
+func eliminatePhysicalProjection(p Plan, replace map[string]*expression.Column) Plan {
+	children := make([]Plan, 0, len(p.Children()))
+	for _, child := range p.Children() {
+		children = append(children, eliminatePhysicalProjection(child, replace))
+	}
+	p.SetChildren(children...)
+
+	for _, dst := range p.Schema().Columns {
+		resolveColumnAndReplace(dst, replace)
+	}
+	for _, key := range p.Schema().Keys {
+		for _, keyCol := range key {
+			resolveColumnAndReplace(keyCol, replace)
+		}
+	}
+	p.replaceExprColumns(replace)
+
+	proj, isProj := p.(*Projection)
+	if !isProj || !canProjectionBeEliminatedStrict(proj) {
+		return p
+	}
+	child := p.Children()[0]
+	exprs := proj.Exprs
+	childCols := child.Schema().Columns
+	for i, parentColumn := range proj.Schema().Columns {
+		col, _ := exprs[i].(*expression.Column)
+		col.DBName = parentColumn.DBName
+		col.TblName = parentColumn.TblName
+		col.ColName = parentColumn.ColName
+		childCols[i].ColName = parentColumn.ColName
+		replace[string(parentColumn.HashCode())] = col
+	}
+	RemovePlan(p)
+	return child
+}
+
+type projectionEliminater struct {
+}
+
+// optimize implements the logicalOptRule interface
+func (pe *projectionEliminater) optimize(lp LogicalPlan, _ context.Context, _ *idAllocator) (LogicalPlan, error) {
+	root := pe.eliminate(lp, make(map[string]*expression.Column), false)
+	return root.(LogicalPlan), nil
+}
+
+// eliminate eliminates the redundant projection in a logical plan.
+// The order of output plan's schema columns can be changed if "haveProjection" is true.
+func (pe *projectionEliminater) eliminate(p Plan, replace map[string]*expression.Column, haveProjection bool) Plan {
+	proj, isProj := p.(*Projection)
+	children := make([]Plan, 0, len(p.Children()))
+	_, isUnion := p.(*Union)
+	for _, child := range p.Children() {
+		children = append(children, pe.eliminate(child, replace, !isUnion && (haveProjection || isProj)))
+	}
+	p.SetChildren(children...)
+
+	_, isSort := p.(*Sort)
+	_, isTopN := p.(*TopN)
+	_, isLimit := p.(*Limit)
+	_, isSelect := p.(*Selection)
+	_, isMaxOneRow := p.(*MaxOneRow)
+
+	_, isApply := p.(*LogicalApply)
+	_, isJoin := p.(*LogicalJoin)
+	if isSort || isTopN || isLimit || isSelect || isMaxOneRow {
```

Comment:
switch x := p.(type) {case *Sort, *TopN ...:} Make sure you don't forget any type by checking this file : https://github.com/pingcap/tidb/blob/master/plan/logical_plans.go